cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+    }
+    const auto diff = cmp(r1.end()->value(), r2.end()->value());
+    if (diff < 0) { // r1 end is strictly before r2 end.
+        return true;
+    }
+    if (diff > 0) { // r1 end is strictly after r2 end.
+        return false;
+    }
+    const auto len1 = r1.end()->value().representation().size();
+    const auto len2 = r2.end()->value().representation().size();
+    if (len1 == len2) { // The values truly are equal.
+        return !r1.end()->is_inclusive() || r2.end()->is_inclusive();
+    } else if (len1 < len2) { // r1 end is a prefix of r2 end.
+        // (a)<(1) ends before (a,b)<=(1,1), but (a)<=(1) doesn't.
+        return !r1.end()->is_inclusive();
+    } else { // r2 end is a prefix of r1 end.
+        // (a,b)<=(1,1) ends before (a)<=(1) but after (a)<(1).
+        return r2.end()->is_inclusive();
+    }
+}
+
+/// Correct clustering_range intersection.  See #8157.
+std::optional<query::clustering_range> intersection(
+        const query::clustering_range& r1,
+        const query::clustering_range& r2,
+        const clustering_key_prefix::prefix_equal_tri_compare& cmp) {
+    // Assume r1's start is to the left of r2's start.
+    if (starts_before_start(r2, r1, cmp)) {
+        return intersection(r2, r1, cmp);
+    }
+    if (!starts_before_end(r2, r1, cmp)) {
+        return {};
+    }
+    const auto& intersection_start = r2.start();
+    const auto& intersection_end = ends_before_end(r1, r2, cmp) ? r1.end() : r2.end();
+    if (intersection_start == intersection_end && intersection_end.has_value()) {
+        return query::clustering_range::make_singular(intersection_end->value());
+    }
+    return query::clustering_range(intersection_start, intersection_end);
+}
+
+/// An expression visitor that translates multi-column atoms into clustering ranges.
+struct multi_column_range_accumulator {
+    const query_options& options;
+    const schema_ptr schema;
+    std::vector<query::clustering_range> ranges{query::clustering_range::make_open_ended_both_sides()};
+    const clustering_key_prefix::prefix_equal_tri_compare prefix3cmp = get_unreversed_tri_compare(*schema);
+
+    void operator()(const binary_operator& binop) {
+        if (auto tup = dynamic_pointer_cast<tuples::value>(binop.rhs->bind(options))) {
+            if (!is_compare(binop.op)) {
+                on_internal_error(
+                        rlogger, format("multi_column_range_accumulator: unexpected atom {}", binop));
+            }
+            auto opt_values = tup->get_elements();
+            auto& lhs = std::get<std::vector<column_value>>(binop.lhs);
+            std::vector<bytes> values(lhs.size());
+            for (size_t i = 0; i < lhs.size(); ++i) {
+                values[i] = deref_column_value(opt_values[i], lhs.at(i).col->name_as_text());
+            }
+            intersect_all(to_range(binop.op, clustering_key_prefix(values)));
+        } else if (auto dv = dynamic_pointer_cast<lists::delayed_value>(binop.rhs)) {
```

Comment:
I think you're relying on rhs->bind() returning `this`. Isn't it more correct to bind once at the top of the function, and check the returned value in each branch?

Reply:
I don't understand what you're suggesting.  Note that this branch never binds the RHS; it calls get_elements() on it.

Reply:
This no longer applies after 1c3de81.